File: day1_2.js
const fs = require("fs");
const assert = require("assert");

const input = fs
  .readFileSync(`./day1.txt`, "utf8")
  .trim()
  .split("\n")
  .map(v => parseInt(v, 10));

// Solution

const requiredFuel = mass => Math.floor(mass / 3) - 2;
const sum = (acc, v) => acc + v;
const recur = (mass, total) => {
  const fuel = requiredFuel(mass);
  return fuel <= 0 ? total : recur(fuel, total + fuel);
};
const solve = i => i.map(v => recur(v, 0)).reduce(sum, 0);

// Tests

assert.strictEqual(solve([14]), 2);
assert.strictEqual(solve([1969]), 966);
assert.strictEqual(solve([100756]), 50346);

console.log("Solution:", solve(input)); // 4878818

File: day1_2.exs
defmodule Starship do
  def required_fuel(mass) do
    Integer.floor_div(mass, 3) - 2
  end

  def fuel_correction(mass) do
    fuel_correction(mass, mass)
  end

  def fuel_correction(mass, total) do
    fuel = required_fuel(mass)
    if fuel <= 0, do: total, else: fuel_correction(fuel, total + fuel)
  end
end

File.read!("./day1.txt")
|> String.split("\n", trim: true)
|> Enum.map(&String.to_integer/1)
|> Enum.map(&Starship.required_fuel/1)
|> Enum.map(&Starship.fuel_correction/1)
|> Enum.sum()
|> IO.puts()
